pkg/resources/kafka/kafka.go

```diff
@@ -877,18 +919,23 @@ func (r *Reconciler) handleRollingUpgrade(log logr.Logger, desiredPod, currentPo
 			if err != nil {
 				return errors.WrapIf(err, "failed to reconcile resource")
 			}
-			for _, pod := range podList.Items {
-				pod := pod
-				if k8sutil.IsMarkedForDeletion(pod.ObjectMeta) {
-					return errorfactory.New(errorfactory.ReconcileRollingUpgrade{}, errors.New("pod is still terminating"), "rolling upgrade in progress")
-				}
-				if k8sutil.IsPodContainsPendingContainer(&pod) {
-					return errorfactory.New(errorfactory.ReconcileRollingUpgrade{}, errors.New("pod is still creating"), "rolling upgrade in progress")
-				}
+			if len(podList.Items) < len(r.KafkaCluster.Spec.Brokers) {
```

Comment:
What's the purpose of this new validation? What edge case does it cover?

Reply:
That is a good question! I was thinking about the following situation:
- failure threshold is 1, concurrent broker restarts is 1 (so the defaults)
- broker 101 was deleted and terminated, it no longer exists
- reconcile runs with current pod 102
- `r.Client.List` returns pods 201, 202, 301, 302 (101 is not returned as it is not created back yet)
- none of pods 201, 202, 301, 302 is either terminating or pending

In this case:
- the actual behavior is that 102 is deleted
- the expected behavior is that we must wait for 101 to be re-created before 102, as only checking for terminating/pending pods is not enough to guarantee there is no broker missing

This seemed to me like an edge case, in the context that all pods are being reconciled in a loop, with 1 pod reconciliation handling either delete (see `reconcileKafkaPod` 2nd switch case, which delegates to `handleRollingUpgrade` which only performs delete), either create (see `reconcileKafkaPod` 1st switch case).
So I enforced any evaluation of pod status or replica availability to happen only if the pod list result has the same amount of pods as the spec.

Reply:
Thanks for the explanation! all clear now